libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/routefetch/RoutesViewModel.kt

```diff
@@ -11,159 +10,94 @@ import com.mapbox.navigation.base.route.NavigationRouterCallback
-import com.mapbox.navigation.dropin.component.destination.DestinationViewModel
-import com.mapbox.navigation.dropin.component.location.LocationViewModel
-import com.mapbox.navigation.dropin.component.navigation.NavigationStateViewModel
-import com.mapbox.navigation.dropin.component.navigationstate.NavigationState
+import com.mapbox.navigation.dropin.extensions.flowRoutesUpdated
 import com.mapbox.navigation.dropin.lifecycle.UIViewModel
-import com.mapbox.navigation.dropin.model.Destination
 import kotlinx.coroutines.flow.collect
-import kotlinx.coroutines.flow.map
 import kotlinx.coroutines.launch
 
 @OptIn(ExperimentalPreviewMapboxNavigationAPI::class)
-internal class RoutesViewModel(
-    private val navigationStateViewModel: NavigationStateViewModel,
-    private val locationViewModel: LocationViewModel,
-    private val destinationViewModel: DestinationViewModel,
-    initialState: RoutesState = RoutesState()
-) : UIViewModel<RoutesState, RoutesAction>(initialState) {
+internal class RoutesViewModel : UIViewModel<RoutesState, RoutesAction>(RoutesState.Empty) {
 
-    private var routeRequestId: Long? = null
+    override fun onAttached(mapboxNavigation: MapboxNavigation) {
+        super.onAttached(mapboxNavigation)
+
+        mainJobControl.scope.launch {
+            mapboxNavigation.flowRoutesUpdated().collect { result ->
+                // Empty is ignored on purpose. When the action is processed
+                // it will be converted to RoutesState.Empty.
+                invoke(RoutesAction.Ready(result.navigationRoutes))
+            }
+        }
+    }
 
     override fun process(
         mapboxNavigation: MapboxNavigation,
         state: RoutesState,
         action: RoutesAction
     ): RoutesState {
-        when (action) {
-            is RoutesAction.FetchAndSetRoute -> {
-                val destination = destinationViewModel.state.value.destination
-                fetchAndSetRoute(mapboxNavigation, destination)
-            }
+        return when (action) {
             is RoutesAction.FetchPoints -> {
                 val routeOptions = getDefaultOptions(mapboxNavigation, action.points)
-                fetchRoute(routeOptions, mapboxNavigation)
+                val requestId = mapboxNavigation.fetchRoute(routeOptions)
+                RoutesState.Fetching(requestId)
             }
             is RoutesAction.FetchOptions -> {
-                fetchRoute(action.options, mapboxNavigation)
+                val requestId = mapboxNavigation.fetchRoute(action.options)
+                RoutesState.Fetching(requestId)
             }
             is RoutesAction.SetRoutes -> {
-                mapboxNavigation.setNavigationRoutes(action.routes, action.legIndex)
+                mapboxNavigation.setNavigationRoutes(action.routes)
+                if (action.routes.isEmpty()) {
+                    RoutesState.Empty
+                } else {
+                    RoutesState.Ready(action.routes)
+                }
```

Comment:
Do we need this? When we set routes on `mapboxNavigation`, `mapboxNavigation.flowRoutesUpdated().collect { result` inside `onAttached` above in this file will trigger `RoutesAction.Ready` which returns the same state as this if/else branch does. Correct me if I am wrong.

Reply:
``` kotlin 
mapboxNavigation.flowRoutesUpdated().collect { result ->
    // Empty is ignored on purpose. When the action is processed
    // it will be converted to RoutesState.Empty.
    invoke(RoutesAction.Ready(result.navigationRoutes))
}
```
The comment is attempting to hint at this possible confusion. There is `RoutesAction.Ready` but not a `RoutesAction.Empty`. 

The conversion to `RoutesState.Empty` needs to happen in the `process`. Also, we want to include a case for every `Action`

![Screen Shot 2022-03-23 at 4 11 14 PM](https://user-images.githubusercontent.com/3021882/159811320-c5f773d9-e2e0-41ad-8b5a-e4fcbea0d0c2.png)

Reply:
Discussed some more here, the redundant logic is because we're exposing a redundant API.

There are two ways to set a route, `MapboxNavigation.setRoutes` and `RoutesViewModel.invoke(RoutesAction.SetRoutes(..)`

Both, will do the exact same thing. But we are using `RoutesViewModel` inside drop-in ui.